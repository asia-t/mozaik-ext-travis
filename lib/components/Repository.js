'use strict';

exports.__esModule = true;

var _templateObject = _taggedTemplateLiteralLoose(['\n    height: 100%;\n    display: flex;\n    flex-direction: column;\n'], ['\n    height: 100%;\n    display: flex;\n    flex-direction: column;\n']),
    _templateObject2 = _taggedTemplateLiteralLoose(['\n    flex: 1;\n    display: flex;\n    align-items: center;\n    ', ';\n'], ['\n    flex: 1;\n    display: flex;\n    align-items: center;\n    ', ';\n']),
    _templateObject3 = _taggedTemplateLiteralLoose(['\n    display: grid;\n    grid-row-gap: 1.6vmin;\n'], ['\n    display: grid;\n    grid-row-gap: 1.6vmin;\n']);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _moment = require('moment');

var _moment2 = _interopRequireDefault(_moment);

var _styledComponents = require('styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _ui = require('@mozaik/ui');

var _state = require('../lib/state');

var _duration = require('../lib/duration');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

function _taggedTemplateLiteralLoose(strings, raw) { strings.raw = raw; return strings; }

var Container = _styledComponents2.default.div(_templateObject);

var Description = _styledComponents2.default.div(_templateObject2, function (props) {
    return (0, _ui.typography)(props.theme, 'default', 'small');
});

var Grid = _styledComponents2.default.div(_templateObject3);

var Repository = function (_Component) {
    _inherits(Repository, _Component);

    function Repository() {
        _classCallCheck(this, Repository);

        return _possibleConstructorReturn(this, _Component.apply(this, arguments));
    }

    Repository.getApiRequest = function getApiRequest(_ref) {
        var owner = _ref.owner,
            repository = _ref.repository;

        return {
            id: 'travis.repository.' + owner + '.' + repository,
            params: { owner: owner, repository: repository }
        };
    };

    Repository.prototype.render = function render() {
        var _props = this.props,
            owner = _props.owner,
            repository = _props.repository,
            title = _props.title,
            repoInfo = _props.apiData,
            apiError = _props.apiError,
            theme = _props.theme;


        var body = _react2.default.createElement(_ui.WidgetLoader, null);
        if (repoInfo) {
            var Icon = (0, _state.iconByState)(repoInfo.last_build.state);
            var color = (0, _state.colorByState)(theme.colors, repoInfo.last_build.state);

            body = _react2.default.createElement(
                Container,
                null,
                _react2.default.createElement(
                    Description,
                    null,
                    repoInfo.description
                ),
                _react2.default.createElement(
                    Grid,
                    null,
                    _react2.default.createElement(_ui.WidgetLabel, {
                        label: _react2.default.createElement(
                            _ui.Text,
                            {
                                variant: 'strong',
                                style: { color: theme.colors.textHighlight }
                            },
                            repoInfo.default_branch.name
                        ),
                        prefix: _react2.default.createElement(_ui.GitBranchIcon, { size: '1.8vmin' })
                    }),
                    _react2.default.createElement(_ui.WidgetLabel, {
                        label: repoInfo.last_build ? _react2.default.createElement(
                            _react.Fragment,
                            null,
                            'last build',
                            ' ',
                            _react2.default.createElement(
                                _ui.Text,
                                { variant: 'strong', style: { color: color } },
                                repoInfo.last_build.state
                            )
                        ) : 'n/a',
                        prefix: _react2.default.createElement(Icon, { size: '1.8vmin', style: { color: color } })
                    }),
                    _react2.default.createElement(_ui.WidgetLabel, {
                        label: repoInfo.last_build ? _react2.default.createElement(
                            'span',
                            null,
                            (0, _moment2.default)(repoInfo.last_build.started_at).fromNow(),
                            ' in',
                            ' ',
                            (0, _duration.secondsToString)(repoInfo.last_build.duration)
                        ) : 'n/a',
                        prefix: _react2.default.createElement(_ui.ClockIcon, { size: '1.8vmin' })
                    }),
                    _react2.default.createElement(_ui.WidgetLabel, {
                        label: 'language',
                        prefix: _react2.default.createElement(_ui.CodeIcon, { size: '1.8vmin' }),
                        suffix: repoInfo.github_language ? repoInfo.github_language : 'n/a'
                    })
                )
            );
        }

        return _react2.default.createElement(
            _ui.Widget,
            null,
            _react2.default.createElement(_ui.WidgetHeader, {
                title: title || '',
                subject: title ? null : owner + '/' + repository,
                icon: _ui.InfoIcon
            }),
            _react2.default.createElement(
                _ui.WidgetBody,
                null,
                _react2.default.createElement(
                    _ui.TrapApiError,
                    { error: apiError },
                    body
                )
            )
        );
    };

    return Repository;
}(_react.Component);

Repository.propTypes = {
    owner: _propTypes2.default.string.isRequired,
    repository: _propTypes2.default.string.isRequired,
    title: _propTypes2.default.string,
    apiError: _propTypes2.default.object,
    apiData: _propTypes2.default.shape({
        slug: _propTypes2.default.string.isRequired,
        description: _propTypes2.default.string.isRequired,
        github_language: _propTypes2.default.string,
        default_branch: _propTypes2.default.shape({
            name: _propTypes2.default.string.isRequired
        }).isRequired,
        last_build: _propTypes2.default.shape({
            state: _propTypes2.default.string.isRequired,
            started_at: _propTypes2.default.string,
            duration: _propTypes2.default.number
        })
    }),
    theme: _propTypes2.default.object.isRequired
};
exports.default = Repository;