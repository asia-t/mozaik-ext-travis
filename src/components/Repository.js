import React, { Fragment, Component } from 'react'
import PropTypes from 'prop-types'
import moment from 'moment'
import styled from 'styled-components'
import {
    TrapApiError,
    Widget,
    WidgetHeader,
    WidgetBody,
    WidgetLoader,
    WidgetLabel as Label,
    InfoIcon,
    ClockIcon,
    GitBranchIcon,
    CodeIcon,
    Text,
    typography,
} from '@mozaik/ui'
import { colorByState, iconByState } from '../lib/state'
import { secondsToString } from '../lib/duration'

const Container = styled.div`
    height: 100%;
    display: flex;
    flex-direction: column;
`

const Description = styled.div`
    flex: 1;
    display: flex;
    align-items: center;
    ${props => typography(props.theme, 'default', 'small')};
`

const Grid = styled.div`
    display: grid;
    grid-row-gap: 1.6vmin;
`

export default class Repository extends Component {
    static propTypes = {
        owner: PropTypes.string.isRequired,
        repository: PropTypes.string.isRequired,
        title: PropTypes.string,
        apiError: PropTypes.object,
        apiData: PropTypes.shape({
            slug: PropTypes.string.isRequired,
            description: PropTypes.string.isRequired,
            github_language: PropTypes.string,
            default_branch: PropTypes.shape({
                name: PropTypes.string.isRequired,
            }).isRequired,
            last_build: PropTypes.shape({
                state: PropTypes.string.isRequired,
                started_at: PropTypes.string,
                duration: PropTypes.number,
            }),
        }),
        theme: PropTypes.object.isRequired,
    }

    static getApiRequest({ owner, repository }) {
        return {
            id: `travis.repository.${owner}.${repository}`,
            params: { owner, repository },
        }
    }

    render() {
        const { owner, repository, title, apiData: repoInfo, apiError, theme } = this.props

        let body = <WidgetLoader />
        if (repoInfo) {
            const Icon = iconByState(repoInfo.last_build.state)
            const color = colorByState(theme.colors, repoInfo.last_build.state)

            body = (
                <Container>
                    <Description>{repoInfo.description}</Description>
                    <Grid>
                        <Label
                            label={
                                <Text
                                    variant="strong"
                                    style={{ color: theme.colors.textHighlight }}
                                >
                                    {repoInfo.default_branch.name}
                                </Text>
                            }
                            prefix={<GitBranchIcon size="1.8vmin" />}
                        />
                        <Label
                            label={
                                repoInfo.last_build ? (
                                    <Fragment>
                                        last build{' '}
                                        <Text variant="strong" style={{ color }}>
                                            {repoInfo.last_build.state}
                                        </Text>
                                    </Fragment>
                                ) : (
                                    'n/a'
                                )
                            }
                            prefix={<Icon size="1.8vmin" style={{ color }} />}
                        />
                        <Label
                            label={
                                repoInfo.last_build ? (
                                    <span>
                                        {moment(repoInfo.last_build.started_at).fromNow()} in{' '}
                                        {secondsToString(repoInfo.last_build.duration)}
                                    </span>
                                ) : (
                                    'n/a'
                                )
                            }
                            prefix={<ClockIcon size="1.8vmin" />}
                        />
                        <Label
                            label="language"
                            prefix={<CodeIcon size="1.8vmin" />}
                            suffix={repoInfo.github_language ? repoInfo.github_language : 'n/a'}
                        />
                    </Grid>
                </Container>
            )
        }

        return (
            <Widget>
                <WidgetHeader
                    title={title || ''}
                    subject={title ? null : `${owner}/${repository}`}
                    icon={InfoIcon}
                />
                <WidgetBody>
                    <TrapApiError error={apiError}>{body}</TrapApiError>
                </WidgetBody>
            </Widget>
        )
    }
}
