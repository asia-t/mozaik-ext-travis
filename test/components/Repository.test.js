import React from 'react'
import { shallow } from 'enzyme'
import renderer from 'react-test-renderer'
import { ThemeProvider } from 'styled-components'
import { WidgetLoader, WidgetHeader, defaultTheme } from '@mozaik/ui'
import Repository from '../../src/components/Repository'

const sampleOwner = 'plouc'
const sampleRepository = 'mozaik'

it('should return correct api request', () => {
    expect(
        Repository.getApiRequest({
            owner: sampleOwner,
            repository: sampleRepository,
        })
    ).toEqual({
        id: `travis.repository.${sampleOwner}.${sampleRepository}`,
        params: {
            owner: sampleOwner,
            repository: sampleRepository,
        },
    })
})

it('should display loader if no apiData available', () => {
    const wrapper = shallow(
        <Repository owner={sampleOwner} repository={sampleRepository} theme={defaultTheme} />,
        {
            context: { theme: {} },
        }
    )

    expect(wrapper.find(WidgetLoader)).toHaveLength(1)
})

it('should display owner/repo', () => {
    const wrapper = shallow(
        <Repository owner={sampleOwner} repository={sampleRepository} theme={defaultTheme} />,
        {
            context: { theme: {} },
        }
    )

    const header = wrapper.find(WidgetHeader)
    expect(header).toHaveLength(1)
    expect(header.prop('title')).toBe('')
    expect(header.prop('subject')).toBe(`${sampleOwner}/${sampleRepository}`)
})

it('should allow title override', () => {
    const wrapper = shallow(
        <Repository
            owner={sampleOwner}
            repository={sampleRepository}
            title="override"
            theme={defaultTheme}
        />,
        { context: { theme: {} } }
    )

    const header = wrapper.find(WidgetHeader)
    expect(header).toHaveLength(1)
    expect(header.prop('title')).toBe('override')
    expect(header.prop('subject')).toBe(null)
})

it('should display info if apiData is available', () => {
    const apiData = {
        id: 3637580,
        slug: 'plouc/mozaik',
        description:
            'Moza√Øk is a tool based on nodejs / reactjs to easily build beautiful dashboards',
        github_language: 'JavaScript',
        default_branch: {
            name: 'master',
        },
        last_build: {
            state: 'passed',
            number: '6',
            duration: 53,
            started_at: '2014-12-29T11:33:09Z',
            finished_at: '2014-12-29T11:34:02Z',
        },
    }

    const tree = renderer.create(
        <ThemeProvider theme={defaultTheme}>
            <Repository
                owner={sampleOwner}
                repository={sampleRepository}
                theme={defaultTheme}
                apiData={apiData}
            />
        </ThemeProvider>
    )

    expect(tree).toMatchSnapshot()
})
