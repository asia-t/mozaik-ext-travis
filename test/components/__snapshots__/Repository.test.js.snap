// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display info if apiData is available 1`] = `
<div
  className="sc-bdVaJa eydJRb"
  style={Object {}}
>
  <div
    className="sc-bwzfXH bJUZVN"
  >
    <header
      className="sc-htpNat lcyfRu"
      style={Object {}}
    >
      <span>
        <div
          className="sc-ifAKCX rpFvk"
        >
          plouc/mozaik
        </div>
        
      </span>
      <div
        className="sc-EHOje iJdEwV"
      >
        <svg
          fill="none"
          height="2vmin"
          stroke="#000"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth="2"
          viewBox="0 0 24 24"
          width="2vmin"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="12"
            cy="12"
            r="10"
          />
          <line
            x1="12"
            x2="12"
            y1="16"
            y2="12"
          />
          <line
            x1="12"
            x2="12"
            y1="8"
            y2="8"
          />
        </svg>
      </div>
    </header>
    <div
      className="sc-bZQynM kFUlyb"
      style={Object {}}
    >
      <div
        className="sc-bRBYWo ldBljU"
      >
        <div
          className="sc-hzDkRC gfEKNv"
        >
          Moza√Øk is a tool based on nodejs / reactjs to easily build beautiful dashboards
        </div>
        <div
          className="sc-jhAzac jjllMB"
        >
          <span
            className="sc-VigVT eprixE"
            style={Object {}}
          >
            <span
              className="sc-fjdhpX kbqJsB"
            >
              <svg
                fill="none"
                height="1.8vmin"
                stroke="currentColor"
                strokeLinecap="round"
                strokeLinejoin="round"
                strokeWidth="2"
                viewBox="0 0 24 24"
                width="1.8vmin"
                xmlns="http://www.w3.org/2000/svg"
              >
                <line
                  x1="6"
                  x2="6"
                  y1="3"
                  y2="15"
                />
                <circle
                  cx="18"
                  cy="6"
                  r="3"
                />
                <circle
                  cx="6"
                  cy="18"
                  r="3"
                />
                <path
                  d="M18 9a9 9 0 0 1-9 9"
                />
              </svg>
            </span>
            <span
              className="sc-jTzLTM itVkgL"
            >
              <span
                className="sc-fBuWsC PbwII"
                style={
                  Object {
                    "color": "#000",
                  }
                }
                type="default"
              >
                master
              </span>
            </span>
          </span>
          <span
            className="sc-VigVT eprixE"
            style={Object {}}
          >
            <span
              className="sc-fjdhpX kbqJsB"
            >
              <svg
                fill="none"
                height="1.8vmin"
                stroke="currentColor"
                strokeLinecap="round"
                strokeLinejoin="round"
                strokeWidth="2"
                style={
                  Object {
                    "color": "#8ddb8d",
                  }
                }
                viewBox="0 0 24 24"
                width="1.8vmin"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M22 11.08V12a10 10 0 1 1-5.93-9.14"
                />
                <polyline
                  points="22 4 12 14.01 9 11.01"
                />
              </svg>
            </span>
            <span
              className="sc-jTzLTM itVkgL"
            >
              last build
               
              <span
                className="sc-fMiknA boQYET"
                style={
                  Object {
                    "color": "#8ddb8d",
                  }
                }
                type="default"
              >
                passed
              </span>
            </span>
          </span>
          <span
            className="sc-VigVT eprixE"
            style={Object {}}
          >
            <span
              className="sc-fjdhpX kbqJsB"
            >
              <svg
                fill="none"
                height="1.8vmin"
                stroke="currentColor"
                strokeLinecap="round"
                strokeLinejoin="round"
                strokeWidth="2"
                viewBox="0 0 24 24"
                width="1.8vmin"
                xmlns="http://www.w3.org/2000/svg"
              >
                <circle
                  cx="12"
                  cy="12"
                  r="10"
                />
                <polyline
                  points="12 6 12 12 16 14"
                />
              </svg>
            </span>
            <span
              className="sc-jTzLTM itVkgL"
            >
              <span>
                4 years ago
                 in
                 
                53 sec
              </span>
            </span>
          </span>
          <span
            className="sc-VigVT eprixE"
            style={Object {}}
          >
            <span
              className="sc-fjdhpX kbqJsB"
            >
              <svg
                fill="none"
                height="1.8vmin"
                stroke="currentColor"
                strokeLinecap="round"
                strokeLinejoin="round"
                strokeWidth="2"
                viewBox="0 0 24 24"
                width="1.8vmin"
                xmlns="http://www.w3.org/2000/svg"
              >
                <polyline
                  points="16 18 22 12 16 6"
                />
                <polyline
                  points="8 6 2 12 8 18"
                />
              </svg>
            </span>
            <span
              className="sc-jTzLTM itVkgL"
            >
              language
            </span>
            <span
              className="sc-fjdhpX kbqJsB"
            >
              JavaScript
            </span>
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
